src/test/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProviderConcurrencyTest.java

```diff
@@ -0,0 +1,303 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.snowflake.kafka.connector.internal.streaming;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig;
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.TestUtils;
+import java.util.*;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+import org.junit.*;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.Mockito;
+
+@RunWith(Parameterized.class)
+public class StreamingClientProviderConcurrencyTest {
+  private Map<String, String> clientConfig;
+
+  private StreamingClientProvider streamingClientProvider;
+  private FakeStreamingClientHandler streamingClientHandler;
+  private boolean enableClientOptimization;
+
+  private List<Future<SnowflakeStreamingIngestClient>> getClientFuturesTeardown;
+  private List<Future> closeClientFuturesTeardown;
+
+  @Parameterized.Parameters(name = "enableClientOptimization: {0}")
+  public static Collection<Object[]> input() {
+    return Arrays.asList(new Object[][] {{true}, {false}});
+  }
+
+  public StreamingClientProviderConcurrencyTest(boolean enableClientOptimization) {
+    this.enableClientOptimization = enableClientOptimization;
+  }
+
+  @Before
+  public void setup() {
+    this.clientConfig = TestUtils.getConfForStreaming();
+    this.clientConfig.put(
+        SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
+        this.enableClientOptimization + "");
+
+    this.streamingClientHandler = new FakeStreamingClientHandler();
+    this.streamingClientProvider =
+        StreamingClientProvider.getStreamingClientProviderForTests(
+            this.streamingClientHandler,
+            StreamingClientProvider.buildLoadingCache(this.streamingClientHandler));
+
+    this.getClientFuturesTeardown = new ArrayList<>();
+    this.closeClientFuturesTeardown = new ArrayList<>();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    // clean up all the threads
+    try {
+      for (Future<SnowflakeStreamingIngestClient> getClientFuture : this.getClientFuturesTeardown) {
+        getClientFuture.get().close();
+      }
+      for (Future closeClientFuture : this.closeClientFuturesTeardown) {
+        closeClientFuture.get();
+      }
+    } catch (Exception ex) {
+      throw ex;
```

Comment:
if we simply rethrow the exception why do we catch them?

Reply:
This is an old test. I just renamed it. Github does not recognize that, but anyway, I can try to fix it.